File: Programmers/Lv2/Lv2_쇠막대기.py
def solution(arrangement):
    answer = 0
    stack= []
    for i in range(len(arrangement)):
        if arrangement[i]=='(':     # 留됰?湲곗쓽 ?쒖옉
            stack.append(i)
        elif arrangement[i]==')':   # 留됰?湲?醫낅즺 or ?덉씠?
            if stack[-1]+1==i:      # ?덉씠???寃쎌슦
                stack.pop()
                answer += len(stack)
            else:                   # 留됰?湲곗쓽 醫낅즺??寃쎌슦
                stack.pop()
                answer += 1
    return answer

print(solution(	"()(((()())(())()))(())"))

File: Programmers/Lv2/Lv2_쇠막대기.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <string>
#include <queue>

using namespace std;

int solution(string arrangement) {
	int answer = 0;
	queue<int> q;

	for (int i = 0; i < arrangement.length() - 1; i++) {
		if (arrangement[i] == '(' && arrangement[i + 1] == '(')		//막대기의 시작
			q.push(i);
	}

	char p, n;
	int index, end;
	answer += q.size();

	while (!q.empty()) {
		index = q.front();
		end = 1;
		q.pop();
		p = arrangement[index];

		for (int i = index + 1; i < arrangement.length(); i++) {
			n = arrangement[i];
			if (n == ')')
				end--;
			else
				end++;

			if (n == ')' && p == ')' && end == 0) {		// 막대기의 종료
				break;
			}
			else if (n == ')' && p == '(') { 		//레이저
				answer++;
			}
			p = n;
		}

	}

	return answer;
}